File: C/src/blockchain.c
#include "blockchain.h"

/*-----------------------------------------------------------------*/

Blockchain* blockchain_create(int difficulte){
	Blockchain *bc = malloc(sizeof(struct s_Blockchain));
	bc->nbBlock = 1;
	bc->difficulte = difficulte;
	//Bloc[0] = bloc génésis
	bc->listeBlocks[0] = bloc_create();
	bc->listeBlocks[0]->index = 0;
	bc->listeBlocks[0]->hashBlocPrecedent = NULL;
	bc->listeBlocks[0]->nonce = 0;
	ajouterTransaction(bc->listeBlocks[0]->listeTX, "Genesis", "Genesis", 1);
	return bc;
}

/*-----------------------------------------------------------------*/

void blockchain_delete(ptrBlockchain *bc){
	for(int i = 0; i < (*bc)->nbBlock; i++)
	{
		bloc_delete(&((*bc)->listeBlocks[i]));
		free((*bc)->listeBlocks[i]);
	}
	free(*bc);
}

/*-----------------------------------------------------------------*/

void ajouterBloc(Blockchain *bc, Bloc *b){
	assert(bc->nbBlock+1 < MAX_BLOC);
	bc->listeBlocks[bc->nbBlock] = b;
	bc->listeBlocks[bc->nbBlock]->index = bc->nbBlock;
	bc->nbBlock++;
}

File: JAVA/src/Blockchain.java
/**\file blockchain.h
 * \date mars 2020
 * \brief La structure pour le blockchain
 */

public class Blockchain {
	public static final int MAX_BLOC = 10;
	private int difficulte;//critère de difficulté pour le minage
	private int nbBlock;//nombre de blocks de la chaîne (en comptant le génésis)
	Bloc listeBlocks[];
	
	/** Constructeur de blockchain
	 * @param la difficulte du minage de bloc
	 */
	public Blockchain(int difficulte) {
		nbBlock = 1;
		this.difficulte = difficulte;
		listeBlocks = new Bloc[MAX_BLOC];
		//Bloc[0] = bloc génésis
		listeBlocks[0] = new Bloc();
		listeBlocks[0].setIndex(0);
		listeBlocks[0].setHashBlocPrecedent("");
		listeBlocks[0].setNonce(0);
		listeBlocks[0].getListeTX().ajouterTransaction("Genesis", "Genesis", 1);
	}
	
	/**
	 * Rattache un bloc à la blockchain (celui de fin)
	 * @param b le bloc
	 * A noter que l'index du bloc ajouté est modifié pour correspondre 
	 * à la blockchain
	 */
	public void ajouterBloc(Bloc b) {
		assert(nbBlock+1 < MAX_BLOC);
		listeBlocks[nbBlock] = b;
		listeBlocks[nbBlock].setIndex(nbBlock);
		nbBlock++;
	}
	
	public boolean estValide() {
		return false;
	}

	/**
	 * @return the difficulte
	 */
	public int getDifficulte() {
		return difficulte;
	}

	/**
	 * @return the listeBlocks
	 */
	public Bloc[] getListeBlocks() {
		return listeBlocks;
	}
}
